File: assignment_enes/problem2.cpp
#include <iostream>
#include <queue>
using namespace std;

class Node
{
public:
    int data;
    Node *next;
    Node(int data=0)
    {
        this->data=data;
        next=NULL;
    }
};

class Queue
{
private:
    Node *front_;
    Node *rear_;
public:
    Queue()
    {
        front_=NULL;
        rear_=NULL;
    }
    bool empty()
    {
        return front_ == NULL;
    }
    void push(int data)
    {
        if(empty())
        {
            front_=new Node(data);
            rear_=front_;
        }
        else
        {
            rear_->next=new Node(data);
            rear_=rear_->next;
        }
    }
    void pop()
    {
        int front;
        front_ = front_->next;
    }

    int front()
    {
        return this->front_->data;
    }

    int rear()
    {
        return this->rear_->data;
    }

    void print()
    {
        Node *temp=front_;
        while(temp!=NULL)
        {
            cout<<temp->data<<"->";
            temp=temp->next;
        }
        cout<<"NULL";
    }
};


Queue q;

void add_to_queue(int num) {
    Queue tempQueue;
    bool added = false;

    while (!q.empty()) {
        int current = q.front();
        q.pop();

        if (!added && current > num) {
            tempQueue.push(num);
            added = true;
        }

        tempQueue.push(current);
    }
    if (!added) {
        tempQueue.push(num);
    }
    q = tempQueue;
}



int main()
{
    q.push(4);
    q.push(7);
    q.push(20);
    q.push(40);
    q.print();cout<<endl;

    add_to_queue(6);
    q.print();
    return 0;
}

File: assignment_enes/problem2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    def __init__(self):
        self.front_ = None
        self.rear_ = None

    def empty(self):
        return self.front_ == None

    def push(self, data):
        if self.empty():
            self.front_ = Node(data)
            self.rear_ = self.front_
        else:
            self.rear_.next = Node(data)
            self.rear_ = self.rear_.next

    def pop(self):
        self.front_ = self.front_.next

    def front(self):
        return self.front_.data

    def rear(self):
        return self.rear_.data

    def print(self):
        temp = self.front_
        while temp != None:
            print(temp.data, "->", end=" ")
            temp = temp.next
        print("NULL")

def add_to_queue(q, num):
    temp_queue = Queue()
    added = False

    while not q.empty():
        current = q.front()
        q.pop()

        if not added and current > num:
            temp_queue.push(num)
            added = True

        temp_queue.push(current)

    if not added:
        temp_queue.push(num)

    return temp_queue

def main():
    q = Queue()
    q.push(4)
    q.push(7)
    q.push(20)
    q.push(40)
    q.print()

    q = add_to_queue(q, 6)
    q.print()

if __name__ == "__main__":
    main()
